src/rust/engine/options/src/arg_splitter.rs

```diff
@@ -0,0 +1,94 @@
+    pub unknown_goals: Vec<String>, // Any requested but unknown goals.
+    pub specs: Vec<String>,         // What to run against, e.g. targets or files/dirs.
+    pub passthru: Vec<String>,      // Any remaining args specified after a -- separator.
+}
+
+pub struct ArgSplitter {
+    build_root: PathBuf,
+    known_goal_names: HashSet<String>,
+}
+
+impl ArgSplitter {
+    pub fn new<'a, I: IntoIterator<Item = &'a str>>(
+        build_root: &Path,
+        known_goal_names: I,
+    ) -> ArgSplitter {
+        ArgSplitter {
+            build_root: build_root.to_owned(),
+            known_goal_names: HashSet::from_iter(known_goal_names.into_iter().map(str::to_string)),
+        }
+    }
+
+    pub fn split_args(&self, args: Vec<String>) -> SplitArgs {
+        let mut goals = vec![];
+        let mut unknown_goals: Vec<String> = vec![];
+        let mut specs = vec![];
+        let mut passthru = vec![];
+
+        let mut unconsumed_args = args;
+        unconsumed_args.reverse();
+        // The first arg is the binary name, so skip it.
+        unconsumed_args.pop();
+
+        // Scan the args looking for goals and specs.
+        // The one hard case is a single word like `foo` with no path- or target-like qualities
+        // (e.g., a slash or a colon). It could be a goal, or it could be a top-level directory.
+        // We disambiguate thus: If it is a known goal name, assume the user intended a goal.
+        // Otherwise, if it looks like a path or target, or exists on the filesystem, assume
+        // the user intended a spec, otherwise it's an unknown goal.
+        // TODO: This is probably not good logic, since the CLI behavior will change based on
+        //  changes to plugins (as they can introduce new goals) or changes on the filesystem.
+        //  We might want to deprecate this behavior and consistently assume that these are goals,
+        //  since the user can always add a `./` prefix to override.
+        while let Some(arg) = unconsumed_args.pop() {
+            // Some special flags, such as `-v` and `--help`, are implemented as
+            // goal aliases, so we must check this before checking for any dash prefixes.
+            if self.known_goal_names.contains(&arg) {
+                goals.push(arg);
+            } else if arg == "--" {
+                // Arg is the passthru delimiter.
+                for item in unconsumed_args.drain(..) {
+                    passthru.push(item);
+                }
+                passthru.reverse();
+            } else if !(arg.starts_with("--") || SINGLE_DASH_FLAGS.contains(arg.as_str())) {
+                // This is not a flag, so it must be an unknown goal or a spec (or a negative spec,
+                // which starts with a single dash, and we know is not a single dash flag).
+                if arg.starts_with("-")
+                    || SPEC_RE.is_match(&arg)
+                    || self.build_root.join(Path::new(&arg)).exists()
```

Comment:
This is sync IO. Does sync vs. async matter at this point of the application?

Reply:
I don't think so - the Python equivalent was blocking on the same thing at the same point. And args are split long before we create a scheduler - we have to split them in order to get the option values for setting up the scheduler.